x-pack/plugins/apm/server/routes/agent_explorer/get_agents_items_details.ts

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { QueryDslQueryContainer } from "@elastic/elasticsearch/lib/api/typesWithBodyKey";
+import { AGENT_NAME, AGENT_VERSION, SERVICE_ENVIRONMENT, SERVICE_LANGUAGE_NAME, SERVICE_NAME, SERVICE_NODE_NAME } from "@kbn/apm-plugin/common/elasticsearch_fieldnames";
+import { environmentQuery } from "@kbn/apm-plugin/common/utils/environment_query";
+import { AgentName } from "@kbn/apm-plugin/typings/es_schemas/ui/fields/agent";
+import { ProcessorEvent } from "@kbn/observability-plugin/common/processor_event";
+import { kqlQuery, rangeQuery } from "@kbn/observability-plugin/server/utils/queries";
+import { RandomSampler } from "../../lib/helpers/get_random_sampler";
+import { AgenItemsSetup } from "./get_agents_items";
+
+export function serviceNameQuery(
+  serviceName?: string
+): QueryDslQueryContainer[] {
+  if (!serviceName) {
+    return [];
+  }
+
+  return [{ term: { [SERVICE_NAME]: serviceName } }];
+}
+
+export function agentLanguageQuery(
+  agentName?: string
+): QueryDslQueryContainer[] {
+  if (!agentName) {
+    return [];
+  }
+
+  return [{ term: { [SERVICE_LANGUAGE_NAME]: agentName } }];
+}
+
+interface AggregationParams {
+  environment: string;
+  serviceName?: string;
+  agentLanguage?: string;
+  kuery: string;
+  setup: AgenItemsSetup;
+  maxNumServices: number;
+  start: number;
+  end: number;
+  randomSampler: RandomSampler;
+}
+
+export async function getAgentsDetails({
+  environment,
+  agentLanguage,
+  serviceName,
+  kuery,
+  setup,
+  maxNumServices,
+  start,
+  end,
+  randomSampler,
+}: AggregationParams) {
+  const { apmEventClient } = setup;
+
+  const response = await apmEventClient.search(
+    'get_agent_details',
+    {
+      apm: {
+        events: [ProcessorEvent.metric],
```

Comment:
Don't we need to get data from other events like transactions, errors and spans?

Reply:
@dgieselaar do you think it would make sense to include the other types of events?

Reply:
That depends - we want as little data to aggregate over, but we need to make sure the data we need is present in at least one of the document types. Ideally we can avoid spans and transaction due to the sheer volume of those. I'm pretty sure we don't need spans. If one of transaction, span or app/system metrics have the fields we need, this is good enough. The `exists` queries will make sure we don't include documents that don't have these fields anyway. Does that make sense?

Reply:
@dgieselaar so you're suggesting to add errors and transactions here or not?

Reply:
@gbamparop I'm suggesting to look into what is needed, I don't know off the top of my head. There is an edge case where if a service only reports errors but has disabled system metrics but I think we can ignore that and not include errors.

Reply:
I see, that makes sense